pontoon/administration/management/commands/sync_projects.py

```diff
@@ -0,0 +1,327 @@
+
+
+class Command(BaseCommand):
+    args = '<project_id project_id ...>'
+    help = 'Synchronize database and remote repositories.'
+
+    def log(self, msg, *args, **kwargs):
+        """Log a message to the console."""
+        self.stdout.write(msg.format(*args, **kwargs) + u'\n')
+
+    def info(self, msg, *args, **kwargs):
+        """Log a message to the console if --verbosity=1 or more."""
+        if self.verbosity >= 1:
+            self.log(msg, *args, **kwargs)
+
+    def debug(self, msg, *args, **kwargs):
+        """Log a message to the console if --verbosty=2."""
+        if self.verbosity == 2:
+            self.log(msg, *args, **kwargs)
+
+    def handle(self, *args, **options):
+        self.verbosity = options['verbosity']
+
+        projects = Project.objects.filter(disabled=False)
+        if args:
+            projects = projects.filter(pk__in=args)
+
+        if len(projects) < 1:
+            raise CommandError('No matching projects found.')
+
+        for project in projects:
+            self.handle_project(project)
+
+        # Once we've synced, we can deleted all translations scheduled
+        # for deletion.
+        Translation.deleted_objects.all().delete()
+
+    def handle_project(self, db_project):
+        if not db_project.can_commit:
+            self.log('Skipping project {0}, cannot commit to repository.'
+                     .format(db_project.name))
+            return
+
+        update_from_repository(db_project)
+        vcs_project = VCSProject(db_project)
+        self.update_resources(db_project, vcs_project)
+
+        vcs_entities = self.get_vcs_entities(vcs_project)
+        db_entities = self.get_db_entities(db_project)
+
+        # Find all possible entity keys.
+        entity_keys = set().union(db_entities.keys(), vcs_entities.keys())
+
+        changeset = ChangeSet(db_project, vcs_project)
+        for key in entity_keys:
+            db_entity = db_entities.get(key, None)
+            vcs_entity = vcs_entities.get(key, None)
+            self.handle_entity(changeset, db_project, key, db_entity, vcs_entity)
+
+        # Apply the changeset to the files and then commit them.
+        changeset.execute()
+        self.commit_changes(db_project)
+
+        # Once we've successfully committed, update the last_synced date.
+        db_project.last_synced = timezone.now()
+        db_project.save()
+
+    def handle_entity(self, changeset, db_project, key, db_entity, vcs_entity):
+        key = db_entity.string if db_entity else vcs_entity.string
+        self.debug(u'Handling entity: `{0}`'.format(key))
+
+        # If the VCS no longer has the entity, we remove it from
+        # Pontoon.
+        if vcs_entity is None:
+            if db_entity is None:
+                # This should never happen. What? Hard abort.
+                raise CommandError('No entities found for key {0}'.format(key))
+            else:
+                # VCS no longer has the entity, remove it from Pontoon.
```

Comment:
We should set obsolete=True instead. Removing entity will also remove translations and thus affect translation memory and contribution stats.